### PR TITLE
Implement LazyListBox for a DialogDependantList #4049

### PR DESCRIPTION
**Pull request by ashklianko**

### CODE REVIEW

File: modules/lib/src/main/resources/assets/js/app/dialog/DependantItemsDialog.ts [hunk 1]
```diff
@@ -71,15 +68,26 @@ export abstract class DependantItemsDialog
...
+
+            this.loadDescendants(size, GetDescendantsOfContentsRequest.LOAD_SIZE).then((newItems: ContentSummaryAndCompareStatus[]) => {
+                if (newItems.length > 0) {
+                    this.addDependantItems(newItems);
+                }
+            }).catch(DefaultErrorHandler.handle)
+                .finally(() => {
+                    this.hideLoadMask();
+                });
+        }, 300));
```
Comment (alansemenov) [inline modules/lib/src/main/resources/assets/js/app/dialog/DependantItemsDialog.ts]: Let's move this to a separate method.

File: modules/lib/src/main/resources/assets/js/app/dialog/DependantItemsDialog.ts [hunk 2]
```diff
@@ -71,15 +68,26 @@ export abstract class DependantItemsDialog
...
+            this.showLoadMask();
+
+            this.loadDescendants(size, GetDescendantsOfContentsRequest.LOAD_SIZE).then((newItems: ContentSummaryAndCompareStatus[]) => {
+                if (newItems.length > 0) {
+                    this.addDependantItems(newItems);
+                }
+            }).catch(DefaultErrorHandler.handle)
+                .finally(() => {
+                    this.hideLoadMask();
+                });
```
Comment (alansemenov) [inline modules/lib/src/main/resources/assets/js/app/dialog/DependantItemsDialog.ts]: No need for curly brackets here

`.finally(() => this.hideLoadMask());`

**Review by alansemenov — CHANGES_REQUESTED** [review]
Instead of having `scrollContainer` as a constructor argument in `LazyListBox` I suggest setting it via setter. Then we won't have to send in `this.getBody()` into every constructor of every descendant of `DialogDependantList` and it will be enough to set it once in `DependantItemsDialog.initElements()`: 

`this.dependantList = this.createDependantList().setScrollContainer(this.getBody());`

**Comment by ashklianko** [comment]
> Instead of having `scrollContainer` as a constructor argument in `LazyListBox` I suggest setting it via setter. Then we won't have to send in `this.getBody()` into every constructor of every descendant of `DialogDependantList` and it will be enough to set it once in `DependantItemsDialog.initElements()`:
> 
> `this.dependantList = this.createDependantList().setScrollContainer(this.getBody());`

`ScrollContainer` is an integral part of `LazyListBox `and must be used immediately to construct `IntersectionObserver`, so it is essential to set it in a constructor, using setter will be more error-prone rather then explicitly setting it in inheritors (imaging if items are set before container is there). One thing I can propose is using Builder class for a `LazyListBox `where constainer could be set together with other observer properties, thus `DialogDependantList `inheritors will need to extend a builder, then  `createDependantList `will need to return Builder class

**Comment by ashklianko** [comment]
Updated

**Review by alansemenov — APPROVED** [review]
Great stuff

**Pull request merged**